design/one-pager-trait-workload-interaction-mechanism.md

```diff
@@ -0,0 +1,198 @@
+# Traits and workloads interaction mechanism in OAM
+
+* Owner: Ryan Zhang (@ryanzhang-oss)
+* Reviewers: Crossplane Maintainers
+* Status: Draft
+
+## Terminology
+
+* **CRD (Custom Resource Definition)** : A standard Kubernetes Custom Resource Definition
+* **CR (Custom Resource)** : An instance of a Kubernetes type that was defined using a CRD
+* **GVK (Group Version Kind)** : The API Group, Version, and Kind for a type of Kubernetes
+ resource (including CRDs)
+* **Workload child resources** : The Kubernetes resources generated by a workload controller. They
+ should all have a owner reference pointing to the parent workload instance.
+
+## Background
+Traits and workloads are two major types of resources in OAM. Traits usually affect how a
+kubernetes resource operate either directly (through spec change) or indirectly (add ingress or
+sidecar). However, the current OAM implementation does not contain a generic mechanism for traits
+ to locate the corresponding resource to modify. 
+ 
+We will use the following hypothetical OAM application as the baseline to illustrate the problem
+ and our solution.
+ 
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: WorkloadDefinition
+metadata:
+name: mydb.standard.oam.dev
+spec:
+definitionRef:
+  name: mydb.standard.oam.dev
+---
+apiVersion: core.oam.dev/v1alpha2
+kind: TraitDefinition
+metadata:
+name: manualscalertraits.core.oam.dev
+spec:
+definitionRef:
+ name: manualscalertraits.core.oam.dev
+---
+apiVersion: core.oam.dev/v1alpha2
+kind: Component
+metadata:
+name: example-db
+spec:
+workload:
+ apiVersion: standard.oam.dev/v1alpha2
+ kind: Mydb
+ metadata:
+   name: mydb-example
+ spec:
+   containers:
+     - name: mysql
+       image: mysql:latest
+---
+apiVersion: core.oam.dev/v1alpha2
+kind: ApplicationConfiguration
+metadata:
+name: example-appconfig
+spec:
+components:
+ - componentName: example-db      
+   traits:
+     - trait:
+         apiVersion: core.oam.dev/v1alpha2
+         kind: ManualScalerTrait
+         metadata:
+           name:  example-appconfig-trait
+         spec:
+           replicaCount: 3
+```
+
+The problem is two folds
+1. A trait controller needs a way to find the workload that it is applied to. 
+   - In the example, the manualScalar trait needs to know that it is supposed to scale the
+    example-db workload. However, since the applicationConfiguration controller just emit a
+     `ManualScalerTrait` CR, it does not contain any information on the `example-db` workload.
```

Comment:
Currently the AppConfig controller assumes all traits will support a `spec.workloadRef` object, and writes said object to each rendered trait.

https://github.com/crossplane/oam-kubernetes-runtime/blob/6b9e04/pkg/controller/oam/applicationconfiguration/apply.go#L68